dFusion/dFusionSpec.md

```diff
@@ -79,10 +79,93 @@ To summarize, here is a list of state that is stored inside the smart contract:
+4. Verifying batch price and trade execution (zkSnark)
+5. Processing of pending exits and deposits (zkSnark)
+6. Restart with step 0
+
+### Account opening & token registration
+
+
+#### Registering accounts
+The contract specifies a constant number of accounts, `K`, that can be opened by anyone by providing an `accountIndex` such that `1 <= accountIndex <= K`. 
+The account index is referred to, synonomously, as the account ID.
+
+Upon opening an account, the contract verifies that
+    - there is no account currently occupying the requested index and
+    - the sender does not currently occupy an account slot.
+
+On the level of contract storage, these contraints imply a bijective mapping {1, 2, ..., K} <-> addresses.
+
+Note that: Registering accounts by specified index (rather than incrementing) enables the possiblity for accounts to be closed and account slots to be made available.
+
+#### Registering tokens
+
+Token registration is performed by token address.
+Tokens are indexed incrementally (implying that they can never be removed) from 1 to `T` (the maximum number of allowed tokens).
+Only the contract owner may register tokens by address.
+
+**Note that** the contract does not verify the requested token address is actually that of an ERC20 token. This implies that security is enforced by permissions (i.e. only owner may register tokens).
+
+In the event that a *non-token* is registered as a token, the indexing slot would be permanently occupied, but no successful deposits could ever be made. However, if the registered token does implement a transferFrom successful deposits could occur
+
+### Deposits & Withdrawals
+
+#### Depositing token
+
+Deposits are requested by token index and amount.
+The contract verifies, that
+        - the sender has registered an account,
+        - the token index is that of a valid registered token and
+        - the sender has sufficient balance
+
+Token transfers (from sender to contract) are made during the deposit request phase, but the sender's balance is not represented in the `stateRoot` until these requests have been processed.
+
+Sender may not exit their balances untill the stateRoot has been updated with the appropriate deposit requests. However, deposit processing will be made publicially available. That is to say, any end user could potentially process their own deposits.
+
+Upon successful transfer, the deposit is included in the appropriate depositRequest slot and the EVM emit's a "Deposit" event containing the following information
+    - Account ID,
+    - Token Index,
+    - Amount Transferred,
+    - Deposit Slot
+
+Where, deposit slot is deterministically governed the EVM's current block number as the integer division of block number by 20. This allows for asynchronicity so that one knows (after a certain block) that the deposit hash is no longer `active` (i.e. will not change). This is required for the asynchronous handling of in-flight transactions.
```

Comment:
If it can be derived from the current block-number, maybe we don't have to provide the block in the event? This would save gas and remove duplicate data.

Reply:
yes interesting, along with the decoded_event data, my listener also gets block information.

Reply:
However, once we incorporate the max deposits per slot, we wont necessarily be able to use block number anymore.